src/util/netif.cpp

```diff
@@ -0,0 +1,305 @@
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl data of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    // Iterate over messages (each message is a routing table entry).
+    for (size_t msg_pos = 0; msg_pos < buf.size(); ) {
+        Assume((msg_pos + sizeof(rt_msghdr)) <= buf.size());
+        const struct rt_msghdr* rt = (const struct rt_msghdr*)(buf.data() + msg_pos);
+        const size_t next_msg_pos = msg_pos + rt->rtm_msglen;
+        Assume(rt->rtm_msglen >= sizeof(rt_msghdr) && next_msg_pos <= buf.size());
+        // Iterate over addresses within message, get destination and gateway (if present).
+        // Address data starts after header.
+        size_t sa_pos = msg_pos + sizeof(struct rt_msghdr);
+        std::optional<CNetAddr> dst, gateway;
+        for (int i = 0; i < RTAX_MAX; i++) {
+            if (rt->rtm_addrs & (1 << i)) {
+                // 2 is just sa_len + sa_family, the theoretical minimum size of a socket address.
+                Assume((sa_pos + 2) <= next_msg_pos);
+                const struct sockaddr* sa = (const struct sockaddr*)(buf.data() + sa_pos);
+                Assume(sa->sa_len >= 2 && (sa_pos + sa->sa_len) <= next_msg_pos);
+                if (i == RTAX_DST) {
+                    dst = FromSockAddr(sa);
+                } else if (i == RTAX_GATEWAY) {
+                    gateway = FromSockAddr(sa);
+                }
+                // Skip to next address.
+                sa_pos += sa->sa_len;
+            }
+        }
+        // Found default gateway?
+        if (dst && gateway && dst->IsBindAny()) { // Route to 0.0.0.0 or :: ?
+            return *gateway;
+        }
+        // Skip to next message.
+        msg_pos = next_msg_pos;
+    }
+    return std::nullopt;
+}
+
+#else
+
+// Dummy implementation.
+static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    (void)family;
+    return std::nullopt;
+}
+
+#endif
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+
+    sa_family_t family;
+    if (network == NET_IPV4) {
+        family = AF_INET;
+    } else if(network == NET_IPV6) {
+        family = AF_INET6;
+    } else {
+        return std::nullopt;
+    }
+
+    std::optional<CNetAddr> ret = QueryDefaultGatewayImpl(family);
+
+    // It's possible for the default gateway to be 0.0.0.0 or ::0 on at least Windows
+    // for some routing strategies. If so, return as if no default gateway was found.
+    if (ret && !ret->IsBindAny()) {
+        return ret;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::vector<CNetAddr> GetLocalAddresses()
+{
+    std::vector<CNetAddr> addresses;
+#ifdef WIN32
+    char pszHostName[256] = "";
+    if (gethostname(pszHostName, sizeof(pszHostName)) != SOCKET_ERROR)
+    {
```

Comment:
style: `{` should be on the same line as `if` or `for` (also in a few places below).